README
$Id: README,v 1.8 2014/03/23 08:37:47 mathes Exp $

README file for Projekte/RotorControl
-----------------------------------------------------------------------------

This is the README file for the rotator controller project.

Directories:
-------------

- Bootloader : contains the boot loeader for the antenna sensor uC

- Docu : various documentation files

- DisplayUR : parts takes from Ulrich Radig's (http://www.ulrichradig.de)
          PowerSupply project. The display unit presented there seems to be
          quite usable for a simple display here, just software needs to be 
          adopted.

- Eagle : contains drawings and schemes done with the Eagle layout program

- Linux : various test programs, to be run on Linux

- LSM303 : library and example for the Pololu LSM303DLH 6-axis sensor
          breakout board. The boards contains a 3-axis magnetometer and a
          3-axis accelerometer.
          See also:
          - http://www.lipoly.de/index.php?main_page=product_info&products_id=165628
          - data sheet LSM303DLH.pdf
          - LSM303DLH-compass-app-note.pdf
          - LSM303DLH-orangutan-example-cde.zip
          - UM10204.pdf (I2C-bus specification)
          - https://github.com/ryantm/LSM303DLH (Arduino library)

- Test : directory containing various test programs for the 'rotorcontrol'
         hardware environment

Project layout (Rev. 1.1)
-------------------------

- the display of UR's PS1 project will be used to display the current and the
  set value for the azimuth (directory DisplayUR)

- a LSM303DLH accelerometer and compass module will be used to determine the
  antenna direction. It will be coupled to the main electronics via a serial
  interface (coupled via RS-485)

- the main electronics module has the task to
  - receive direction information from the LSM303DLH
  - calculate/average the direction information and send it to the display 
    module
  - compare the current direction with the preset value and pass this
    information to the rotator controller (software module)
  - switch the rotator motor in an appropriate way to achieve the desired
    direction; handle (optional) end-switches and a clutch/brake
  - get users input via push buttons (attached to the main electronics or to
    the display module?)
  - ...

=============================================================================



CHANGES
$Id: CHANGES,v 1.55 2014/04/01 05:37:40 mathes Exp $

CHANGES file for Projekte/RotorControl
-----------------------------------------------------------------------------

ToDo next:
----------
- extend I2C driven display with more possibilities
- use watchdog to reset in case of hang up
  -> then only partly initialisation
- read constants initially from EEPROM


Version xxxx: (not yet tagged)
-------------

2014-03-22 (thjm) - Bootloader directory added

2013-04-06 (thjm) - AntennaBoard: ERC errors fixed, worked on board and DRC
                  - PowerSupply: ERC & DRC errors fixed

Version v1r0:
-------------

2012-06-07 (thjm) - added GS232 protocol definitions to Docu:
                   - GS232A.pdf
		   - Yaesu_GS-232B_Manual.pdf

2012-06-06 (thjm) - actions for 'preset' functionality disabled
                    + catch position always moving

2012-06-05 (thjm) - messages to UART sometimes corrupt, fixed
                    -> UART_TX_BUFFER_SIZE, UART_RX_BUFFER_SIZE in Makefile
		       + different sizes (RX buffer large)
		  - read constants from EEPROM failed with int16_t -> float OK

2012-06-04 (thjm) - heading calculation fixed, first mounting of mockup
                  - first version with 'preset' -> with some problems ...

2012-06-03 (thjm) - heading calculation now included
                    + linking with scanf-library for floating point !?
		    + or using atoi() based string parsing (saves 2.5k)
		    + vector.c and num2uart.c included into Makefile
		  - heading written to display
		  - average heading is also calculated, number of samples
		    can be configured easily...
		  - Makefile: target 'eeprom' added

2012-05-25 (thjm) - AntennaBoard: now with ISP-header, U.Radig layout
                  - Overview.odp: schematic overview of the system

2012-05-24 (thjm) - added datasheet for MAX485 & Co.
                  - AntennaBoard: finalized (?) and first board layout

2012-05-20 (thjm) - copyright stuff added (GNU GPL v3)
                  - code moved to compass.c, untested

2012-05-** (thjm) - PowerSupply: schematic continued, proposal for board layout
                  - i2cdisplay.h & i2cdisplay.c: I2C display routines
		  - start message "DC2IP"
		    -> this message doesn't always appear at start !?
		  - some work already on preset counter
		  - nested state machine, for CW, CCW and STOP command
		  - state machine for preset commands implemented
		  - auto power-off for preset display
		  - function (algorithm) to determine rotation direction
		  - started with some configuration variables in EEPROM

2012-05-14 (thjm) - simple rotator interaction moved to rotorstate.c
                  - get8key4.c recovered, was deleted in LSM303 directory

2012-05-13 (thjm) - cabling of RelayBoard1 partly changed
                  - some pin-swaps to ButtonBoard
		  -> blrtest.c finally works as expected

2012-05-12 (thjm) - RS485_PORT defined separately
                  - CNT0_PRESET recalculated for 12 MHz
		  - swapped LED_L and LED_R in schematic and global.h
		  - solder bridge between two buttons found and fixed
		  - rotorstate.c
		  - Test/uarttest.c, this is for USART0 (the only one on ATmega32)

2012-05-11 (thjm) - directory for test programs: Test
                    - displaytest.c etc. moved here
		  - blrtest.c: test program for buttons, LEDs and relays
		  - additional defines RELAY1 .. 4 introduced
		  - started with AntennaBoard (need drawing for LSM303DLH)

2012-05-10 (thjm) - CPUBoard: fix connections to display, somehow mirrored
                  - wiring from/to MAX232 changed slightly
		  - displaytest.c (from DisplayUR/twitest.c)
		  - after another fix to TWI pins -> displaytest.c runs!!!
		  - swapped connections to relay board1

2012-05-09 (thjm) - ATmega32 put finally on board
                    -> fuse programming (see rotorcontrol.c)
		    -> test with avrledtest (OK)

2012-05-08 (thjm) - CPUBoard:
                    - another pin-header added for RS-485 UART
		    - correct termination for RS-485 A/B lines
		  - PowerSupply: terminals added
		  - eagle_clean.sh: remove backup files of Eagle

2012-05-07 (thjm) - corrections for RelayBoard1:
                    - using right connectors from library con-molex
		    - single sided routing
		  - corrections for ButtonBoard:
		    - using right connector from library con-molex
		    - using right switches from library switch-misc (SRK/SRKL)
		  - component placement & board layout for ButtonBoard done
		  - CPUBoard:
		    - some components moved around the board
		    - re-routing (partially)

2012-05-06 (thjm) - corrections for CPUBoard:
                    - RS-485 need termination resistors etc.
		    - connectors are wrong, use 22-?-05 tec. from con-molex
		  -> connectors on other boards must be replaced, too!

2012-05-04 (thjm) - CPUBoard with some corrections, board routed (demo)

2012-04-27 (thjm) - RelayMountingPlate.jpg, tnx Ira
                  - RelayNaming.odp, includes also wiring scheme

2012-04-19 (thjm) - CPUBoard: jumper block for serial interface, MAX232 | USB

2012-04-18 (thjm) - changed global.h according to recent CPUboard schematic
                  - initialisation code for LEDs and BUTTONs added

2012-04-17 (thjm) - schematic for CPUBoard mostly complete

2012-04-11 (thjm) - ATmega-40pin.txt: overview of different pinouts

2012-04-04 (thjm) - schematic (incomplete) for power supply added

2012-03-30 (thjm) - schematic for relay board renamed
                  - second relay board added
		  - button board added
		  - main board (CPU) started

2012-01-26 (thjm) - circuit for relay board added
                  - pin definitions changed for nicer circuit drawing

2012-01-25 (thjm) - definition for rotor power relay

2012-01-03 (thjm) - Atmega32 will work with 12 MHz crystal

2011-12-29 (thjm) - button control for LEFT/RIGHT added
                  - timer controlled button query integrated
		    -> not yet tested !


global.h

/*
 * File   : global.h
 *
 * $Id: global.h,v 1.26 2014/03/23 07:30:35 mathes Exp $
 *
 * Copyright:      Hermann-Josef Mathes  mailto: dc2ip@darc.de
 * Author:         Hermann-Josef Mathes
 * Remarks:
 * Known problems: development status
 * Version:        $Revision: 1.26 $ $Date: 2014/03/23 07:30:35 $
 * Description:    Contains all global definitions of the 'rotorcontrol'
 *                 project.
 *

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
   If not, write to the Free Software Foundation,
   Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

 *
 */


#ifndef _global_h_
#define _global_h_

/** @file global.h
  * Global header file for 'rotorcontrol' project.
  * @author H.-J. Mathes
  */


#include <stdint.h>
#include <avr/io.h>
#include <avr/eeprom.h>

#ifndef TRUE
 #define TRUE 	(1==1)
#endif /* TRUE */
#ifndef FALSE
 #define FALSE  (1==0)
#endif /* FALSE */

#include "vector.h"

/* --- for the UART library of P.Fleury --- */

//#define UART_TX_BUFFER_SIZE	128
//#define UART_RX_BUFFER_SIZE	128

// ==> uart.h must be included afterwards !!!

/* --- my program constants --- */

// 12 MHz crystal ==> CLK/1024 = 11.71875 kHz
// T_0 = 0.0853 msec ==> * 117 = 9.984 msec = T_1
#define CNT0_PRESET             (0xff - 117)

#define BUTTON_PORT             PORTA
#define BUTTON_PIN              PINA
#define BUTTON_DDR              DDRA

#define BUTTON_PRESET_CCW       (1<<PA4)
#define BUTTON_CCW              (1<<PA2)
#define BUTTON_STOP             (1<<PA0)
#define BUTTON_CW               (1<<PA1)
#define BUTTON_PRESET_CW        (1<<PA3)

#define LED_PORT                PORTC
#define LED_DDR                 DDRC

#define LED_OVERLOAD            (1<<PC2)  // red
#define LED_CALIBRATE           (1<<PC3)  // yellow
#define LED_LEFT                (1<<PC4)
#define LED_RIGHT               (1<<PC5)

#define RS485_PORT              LED_PORT
#define RS485_DDR               LED_DDR

#define RS485_TX_ENABLE         (1<<PC6)

#define RS485EnableRx()         { RS485_PORT &= ~RS485_TX_ENABLE; }
#define RS485EnableTx()         { RS485_PORT |= RS485_TX_ENABLE; }

#define RELAY_PORT              PORTD
#define RELAY_DDR               DDRD

#define RELAY1                  (1<<PD7)  // generic names, as in schematic
#define RELAY2                  (1<<PD6)
#define RELAY3                  (1<<PD5)
#define RELAY4                  (1<<PD4)

#define RELAY_POWER             RELAY4    // turn on rotor power
#define RELAY_STOP              RELAY3    // open the brake
#define RELAY_CCW               RELAY2    // turn counter clock wise
#define RELAY_CW                RELAY1    // turn clock wise

#define BrakeLock()             { RELAY_PORT &= ~RELAY_STOP; }
#define BrakeRelease()          { RELAY_PORT |= RELAY_STOP; }
#define PowerOn()               { RELAY_PORT |= RELAY_POWER; }
#define PowerOff()              { RELAY_PORT &= ~RELAY_POWER; }
#define RotatorCW()             { RELAY_PORT |= RELAY_CW; }
#define RotatorCCW()            { RELAY_PORT |= RELAY_CCW; }
#define RotatorOff()            { RELAY_PORT &= ~(RELAY_CW | RELAY_CCW); }

/** vector like data type with int16_t components. */
typedef struct i_vector {

  int16_t  x;
  int16_t  y;
  int16_t  z;

} i_vector_t;

/* --- configuration data in EEPROM --- */

//extern i_vector_t EEMEM gEE_MAG_min;
//extern i_vector_t EEMEM gEE_MAG_max;
extern vector_t EEMEM gEE_MAG_min;
extern vector_t EEMEM gEE_MAG_max;

/* --- declaration(s) for file get8key4.c --- */

extern volatile uint8_t gKeyState;
extern volatile uint8_t gKeyPress;

extern void CheckKeys(void);
extern uint8_t GetKeyPress(uint8_t key_mask);
extern uint8_t GetKeyShort(uint8_t key_mask);

/* --- declaration(s) for file rotorstate.c --- */

/** Set the current heading (for the display).
  *
  * If we are not in 'preset' mode, the gPresetHeading is also modified.
  */
extern void SetCurrentHeading(int);

/**  */
extern void SetPresetHeading(int);

/**  */
extern void RotatorExec(void);

extern volatile uint8_t gRotatorBusy;
extern volatile uint8_t gRotatorCommand;
extern volatile uint8_t gRotatorState;
extern volatile uint8_t gRotatorBusyCounter;

#define IsRotatorBusy() (gRotatorBusy != 0)

/** States the rotor control might be in. */

typedef enum {

  kIdle,
  kReleaseBrake,
  kLockBrake,
  kRotorRampup,
  kRotorRampdown,
  kTurningCCW,
  kTurningCW,

} ERotorState;

/** The commands to the rotor control. */
typedef enum {

  kNone,
  kStop,
  kTurnCCW,
  kTurnCW,
  kFastStop,

} ERotorCommand;

/**  */
extern void SetCommand(uint8_t cmd);
/**  */
extern uint8_t GetLastCommand(void);

/**  */
extern void SetPresetCommand(uint8_t cmd);

/**  */
extern void PresetExec(void);
extern void PresetExecSlow(void);
/**  */
extern void UpdateDisplay(void);

/** Minimum and maximum angles. */
#define MIN_ANGLE        0
#define MAX_ANGLE      359

/** A mechanical stop where we cannot rotate further. */
#define LIMIT_ANGLE    270

typedef enum {

  kPresetNone = 0,
  kPresetCW,
  kPresetCCW,
  kPresetStop,
  kPresetExec,

} EPresetCommand;

/** Command to the 'preset' logic. */
extern volatile uint8_t gPresetCommand;
extern volatile uint8_t gPresetCounter;

extern volatile uint16_t gPresetDisplayCounter;

/* --- declaration(s) for file compass.c --- */

extern void CompassInit(void);
extern void CompassMessageInit(void);
extern void CompassMessageReceive(unsigned int uart_data);

#endif /* _global_h_ */
